plugins/GSdx/GSRendererHW.cpp

```diff
@@ -138,6 +129,52 @@ void GSRendererHW::SetScaling()
 	printf("Frame buffer size set to  %dx%d (%dx%d)\n", fb_width, fb_height , m_width, m_height);
 }
 
+void GSRendererHW::CustomResolutionScaling()
+{
+	const int crtc_width = GetDisplayRect().width();
+	const int crtc_height = GetDisplayRect().height();
+	const float scaling_ratio = ceil(static_cast<float>(m_custom_height) / crtc_height);
+	// Avoid using a scissor value which is too high, developers can even leave the scissor to max (2047)
+	// at some cases when they don't want to limit the rendering size. Our assumption is that developers
+	// set the scissor to the actual data in the buffer. Let's use the scissoring value only at such cases
+	const int scissor_height = std::min(640u, (m_context->SCISSOR.SCAY1 - m_context->SCISSOR.SCAY0) + 1);
+	const int single_buffer_size = std::max(GetDisplayRect().height(), scissor_height);
+
+	// We have two contexts of framebuffer height -
+	// One for lower memory consumption and low accuracy
+	// Another one for higher memory consumption at necessary scenarios for higher accuracy
+	std::array<int, 2> framebuffer_height;
+	// When Large Framebuffer is disabled - Let's only consider the height of the display rectangle
+	// as the base (This is wrong implementation when we consider it theoretically as CRTC has no relation
+	// to the Framebuffer size)
+	framebuffer_height[0] = static_cast<int>(round((crtc_height * scaling_ratio)));
+	// When Large Framebuffer is enabled - We also consider for potential scissor sizes which are around
+	// the size of the actual image data stored. (Helps ICO to properly scale to right size by help of the
+	// scissoring values) Display rectangle has a height of 256 but scissor has a height of 512 which seems to
+	// be the real buffer size.
+	framebuffer_height[1] = static_cast<int>(round(single_buffer_size * scaling_ratio));
+
+	if (m_width >= m_custom_width && m_height >= framebuffer_height[m_large_framebuffer])
+		return;
+
+	m_tc->RemovePartial();
+	m_width = max(m_width, native_buffer.x);
+	m_height = max(framebuffer_height[m_large_framebuffer], native_buffer.y);
+
+	const int overhead = framebuffer_height[1] - framebuffer_height[0];
+	std::unique_ptr<char[]> buffer = std::make_unique<char[]>(150);
+	int size = snprintf(buffer.get(), 150, "(Custom resolution) Framebuffer size set to %dx%d (%dx%d)\n", crtc_width, crtc_height, m_width, m_height);
```

Comment:
Initially I wanted to do it in C++ style, something like,

``` cpp
std::stringstream s;
s >> "(Custom resolution) Framebuffer size set to" << crtc_width << "x" ...;
```

The thing is that on both this and your solution, the code will be too long horizontally, so I decided to just format it using snprintf. What's wrong with the current way? This is the traditional C way to format strings. (Though aside from using smart pointer instead of bare one)